pkg/git/git.go

```diff
@@ -0,0 +1,118 @@
+func (TheGitUtils) CommitSingleFile(filePath, commitMessage string, worktree UtilsWorkTree) (plumbing.Hash, error) {
+	_, gitAddError := worktree.Add(filePath)
+	if gitAddError != nil {
+		log.Entry().WithError(gitAddError).Error("Failed to add file to git")
+		return [20]byte{}, gitAddError
+	}
+
+	commit, commitError := worktree.Commit(commitMessage, &git.CommitOptions{})
+	if commitError != nil {
+		log.Entry().WithError(commitError).Error("Failed to commit file")
+		return [20]byte{}, commitError
+	}
+
+	return commit, nil
+}
+
+// PushChangesToRepository Pushes all committed changes in the repository to the remote repository
+func (TheGitUtils) PushChangesToRepository(username, password string, repository UtilsRepository) error {
+	pushOptions := &git.PushOptions{
+		Auth: &http.BasicAuth{Username: username, Password: password},
+	}
+	pushError := repository.Push(pushOptions)
+	if pushError != nil {
+		log.Entry().WithError(pushError).Error("Failed to push commit")
+		return pushError
+	}
+	return nil
+}
+
+// PlainClone Clones a non-bare repository to the provided directory
+func (TheGitUtils) PlainClone(username, password, serverURL, directory string) (UtilsRepository, error) {
+	gitCloneOptions := git.CloneOptions{
+		Auth: &http.BasicAuth{Username: username, Password: password},
+		URL:  serverURL,
+	}
+	repository, gitCloneError := abstractedGit.plainClone(directory, false, &gitCloneOptions)
+	if gitCloneError != nil {
+		log.Entry().WithError(gitCloneError).Error("Failed to clone git")
+		return nil, gitCloneError
+	}
+	return repository, nil
+}
+
+// ChangeBranch checkout the provided branch.
+// It will create a new branch if the branch does not exist yet.
+// It will checkout "master" if no branch name if provided
+func (TheGitUtils) ChangeBranch(branchName string, worktree UtilsWorkTree) error {
+	if branchName == "" {
+		branchName = "master"
+	}
+
+	var checkoutOptions = &git.CheckoutOptions{}
+	checkoutOptions.Branch = plumbing.NewBranchReferenceName(branchName)
+	checkoutOptions.Create = false
+	checkoutError := worktree.Checkout(checkoutOptions)
+	if checkoutError != nil {
+		// branch might not exist, try to create branch
+		checkoutOptions.Create = true
+		checkoutError = worktree.Checkout(checkoutOptions)
+		if checkoutError != nil {
+			log.Entry().WithError(checkoutError).Error("Failed to checkout branch")
+			return checkoutError
+		}
+	}
+
+	return nil
+}
+
+// GetWorktree returns the worktree of the repository
+func (TheGitUtils) GetWorktree(repository UtilsRepository) (UtilsWorkTree, error) {
+	worktree, err := repository.Worktree()
+	if err != nil {
+		log.Entry().WithError(err).Error("could not receive worktree from repository")
+	}
+	return worktree, err
+}
+
+type abstractionGit struct{}
+
+func (abstractionGit) plainClone(path string, isBare bool, o *git.CloneOptions) (*git.Repository, error) {
+	return git.PlainClone(path, isBare, o)
+}
```

Comment:
I am not completely sure about the setup of this file. From what I understand, you want to provide functionality where the actual execution is delegated either to a mocking or runtime version of `UtilsWorkTree` and `UtilsRepository `. Why are these interfaces exported? You can also export functions which take an interface that you declare privately. It should be possible to name these interfaces `worktree` and `repository` without them clashing with anything. And why do you need the type `TheGitUtils`? If you had plain functions, you would access them via the package prefix, like `git.CommitSingleFile()`, which seems elegant to me. Could you elaborate on the purpose of `abstractionGit`? It is not clear to me, sorry about that.

Reply:
`UtilsWorkTree`, `UtilsRepository`, `utilsGit` are interface to abstract the git implementation. This is regular way to handle dependencies in a way such that the dependency can be mocked properly (git_test.go) would not be possible without. The same way `gitopsGitUtils` is handled in gitopsUpdateDeployment.go.

If I wouldn't use `TheGitUtils` the methods would be like static without a wrapping util "object". Without that it would not be possible to mock the properly as the interface implementation of `gitopsGitUtils` in gitopsUpdateDeployment.go will need a connection to the implementation.

This brings me to `abstractionGit`, this is exactly what happens when there is no wrapping "object". There is no possibility to mock the plaincall command from git. Hence, I need to create a struct that implements an interface that wraps the PlainClone method to mock it properly in tests.

Reply:
See https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
I frequently point to this location myself. :-)

My point is a different one. Shall we have a call?

Reply:
Resolved after a call.